newsroom/search.py

```diff
@@ -66,17 +66,112 @@ class BaseSearchService(Service):
+            self.prefill_search_query(search, req, lookup)
+            self.validate_request(search)
+            self.apply_filters(search)
+            self.gen_source_from_search(search)
+
+            internal_req = self.get_internal_request(search)
+            return self.internal_get(internal_req, search.lookup)
+
+    def on_fetched(self, docs):
+        """Add IDs of the versions that matched the search to the HATEOAS response
+
+        The list of IDs are not guaranteed to be in the ``_items`` response.
+        This is used in the front-end to highlight which version matched the search query,
+        which may not be the last version in the content chain.
+        """
+
+        if self._matched_ids:
+            docs['_links']['matched_ids'] = self._matched_ids
+            self._matched_ids = []
+
+    def _search_all_versions(self, search: SearchQuery, req, lookup):
+        """Search across all versions of items, but return last versions only"""
+
+        search.args['ignore_latest'] = True
         self.prefill_search_query(search, req, lookup)
         self.validate_request(search)
         self.apply_filters(search)
         self.gen_source_from_search(search)
 
+        # Search up to 1,000 items to make sure pagination works
+        # as we're getting all versions here
+        # where as the final response will only include the last version
+        # of each content chain
+        search.source['size'] = 1000
+        search.source['from'] = 0
+
+        internal_req = self.get_internal_request(search)
+        search_results = self.internal_get(internal_req, search.lookup)
+        next_item_ids = []
+        self._matched_ids = []
+
+        for doc in search_results.docs:
+            self._matched_ids.append(doc['_id'])
+            next_item_ids.append(str(self.get_last_version(doc)['_id']))
+
+        # Now run a query only using the IDs from the above search
+        # This final search makes sure pagination still works
+        search.query['bool'] = {'must': {'terms': {'_id': next_item_ids}}}
+        self.gen_source_from_search(search)
         internal_req = self.get_internal_request(search)
         return self.internal_get(internal_req, search.lookup)
 
+    def get_last_version(self, doc):
+        if not doc.get('nextversion'):
+            # This is already the latest version
+            return doc
+        elif doc.get('original_id'):
+            # Attempt to get the last version in the series using Elastic
+            original_id = doc['original_id']
+            req = ParsedRequest()
+            req.args = {
+                'source': json.dumps({
+                    'query': {
+                        'bool': {
+                            'must': [
+                                {'term': {'original_id': original_id}},
+                            ],
+                            'must_not': [
+                                {'constant_score': {'filter': {'exists': {'field': 'nextversion'}}}}
```

Comment:
just wondering why using the constant score here?

Reply:
Just bad habit I suppose from Elastic 1.x days